src/datasets/arrow_dataset.py

```diff
@@ -354,56 +357,89 @@ def to_tf_dataset(
-        elif isinstance(columns, str):
-            columns = [columns]
-        elif len(set(columns)) < len(columns):
-            raise ValueError("List of columns contains duplicates.")
-        if label_cols is not None:
-            cols_to_retain = columns + label_cols
+        if columns:
+            if isinstance(columns, str):
+                columns = [columns]
+            if len(set(columns)) < len(columns):
+                raise ValueError("List of columns contains duplicates.")
+            cols_to_retain = list(set(columns + label_cols))
         else:
-            cols_to_retain = columns
-        if "label" in cols_to_retain or "labels" in cols_to_retain or "label_ids" in cols_to_retain:
-            cols_to_retain += ["labels", "label", "label_ids"]  # Don't accidentally drop any labels with other names!
-        cols_to_retain = list(set(cols_to_retain))  # Remove any duplicates
-
+            cols_to_retain = None  # Indicates keeping all non-numerical columns
+            columns = []
         if drop_remainder is None:
             # We assume that if you're shuffling it's the train set, so we drop the remainder unless told not to
             drop_remainder = shuffle
+        # endregion
 
-        retained_columns = [key for key in self.features.keys() if key in cols_to_retain]
-        dataset = self.with_format("numpy", columns=retained_columns)
+        # region Set dataset format and drop unwanted
 
-        columns_to_dtypes, output_signature = self._get_output_signature(
-            dataset, collate_fn, collate_fn_args, batch_size=batch_size if drop_remainder else None
+        if self.format["type"] != "custom":
+            dataset = self.with_format("numpy")
+        else:
+            dataset = self
+        if cols_to_retain is not None:
+            # Following the logic in `transformers.Trainer`, we do not drop `label_ids` or `label` even if they
+            # are not in the list of requested columns, because the collator may rename them
+            # This might work better if moved to a method attached to our transformers Model objects, but doing so
+            # could break backward compatibility
+            unwanted_columns = [
+                col
+                for col in dataset.features.keys()
+                if col not in cols_to_retain and col not in ("label_ids", "label")
+            ]
+            dataset = dataset.remove_columns(unwanted_columns)
+        # If the user hasn't specified columns, give them all columns. This may break some data collators if columns
+        # are non-numeric!
+
+        # endregion
+
+        # region Compute dataset signature and verify columns
+        output_signature, columns_to_np_types = dataset._get_output_signature(
+            dataset,
+            collate_fn=collate_fn,
+            collate_fn_args=collate_fn_args,
+            batch_size=batch_size if drop_remainder else None,
```

Comment:
why does `drop_remainder` matter here ?

Reply:
If we disable `drop_remainder` then the dataset will output batches with variable length on the batch dimension. Therefore, the dataset's output shape should display that dimension as `None`. If we do use `drop_remainder` then that dimension will be constant, so we can include it in the dataset output shape.

Reply:
Sure!

Reply:
Done!